File: Algorithms Advanced/others/ShortestPath/DijkstraSP.py
# Dijkstra's Algorithm (Single Source, Non-negative Weights)
import heapq

def dijkstra(source, nodes, adj):
    dist = [float('inf')] * nodes
    dist[source] = 0
    pq = [(0, source)]  # Priority queue as (distance, node)

    while pq:
        current_dist, u = heapq.heappop(pq)

        for v, weight in adj[u]:
            if current_dist + weight < dist[v]:
                dist[v] = current_dist + weight
                heapq.heappush(pq, (dist[v], v))

    for i in range(nodes):
        print(f"Distance from {source} to {i} is {dist[i]}")

# Example
nodes = 5
adj = {
    0: [(1, 10), (4, 3)],
    1: [(2, 2)],
    2: [(3, 9)],
    3: [],
    4: [(1, 1), (2, 8)]
}
dijkstra(0, nodes, adj)

File: Algorithms Advanced/others/ShortestPath/DijkstraSP.cpp
// Dijkstra's Algorithm (Single Source, Non-negative Weights)
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

void dijkstra(int source, int nodes, vector<vector<pair<int, int>>> &adj) 
{
    vector<int> dist(nodes, INT_MAX);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;

    dist[source] = 0;
    pq.push({0, source});

    while (!pq.empty()) 
    {
        int u = pq.top().second;
        pq.pop();

        for (auto &[v, weight] : adj[u]) 
        {
            if (dist[u] + weight < dist[v]) 
            {
                dist[v] = dist[u] + weight;
                pq.push({dist[v], v});
            }
        }
    }

    for (int i = 0; i < nodes; i++) 
    {
        cout << "Distance from " << source << " to " << i << " is " << dist[i] << endl;
    }
}

int main() 
{
    int nodes = 5;
    vector<vector<pair<int, int>>> adj(nodes);
    adj[0] = {{1, 10}, {4, 3}};
    adj[1] = {{2, 2}};
    adj[2] = {{3, 9}};
    adj[3] = {};
    adj[4] = {{1, 1}, {2, 8}};

    dijkstra(0, nodes, adj);
    return 0;
}
